SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc

```diff
@@ -1,461 +1,673 @@
+
+#include "SimDataFormats/CaloAnalysis/interface/CaloParticle.h"
+#include "SimDataFormats/CaloAnalysis/interface/CaloParticleFwd.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimClusterFwd.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+#include "SimDataFormats/CaloTest/interface/HGCalTestNumbering.h"
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/Vertex/interface/SimVertex.h"
+
+#include "SimGeneral/MixingModule/interface/DigiAccumulatorMixMod.h"
+#include "SimGeneral/MixingModule/interface/DigiAccumulatorMixModFactory.h"
+#include "SimGeneral/MixingModule/interface/PileUpEventPrincipal.h"
+#include "SimGeneral/TrackingAnalysis/interface/EncodedTruthId.h"
 
 #include "Geometry/CaloGeometry/interface/CaloGeometry.h"
-#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
 #include "Geometry/HcalCommonData/interface/HcalHitRelabeller.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
 
-#include "SimDataFormats/CaloAnalysis/interface/CaloParticle.h"
-#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+namespace {
+using Index_t = unsigned;
+using Barcode_t = int;
+}
 
-#include <iterator>
+using boost::adjacency_list;
+using boost::directedS;
+using boost::listS;
+using boost::vecS;
+using boost::property;
+using boost::edge;
+using boost::edge_weight_t;
+using boost::edge_weight;
+using boost::add_edge;
+using boost::vertex;
+using boost::vertex_name_t;
+using boost::vertex_name;
+
+/* GRAPH DEFINITIONS
+
+   The graph represents the full decay chain.
+
+   The parent-child relationship is the natural one, following "time".
+
+   Each edge has a property (edge_weight_t) that holds a const pointer to the
+   SimTrack that connects the 2 vertices of the edge, the number of simHits
+   associated to that simTrack and the cumulative number of simHits of itself
+   and of all its children. Only simHits within the selected detectors are
+   taken into account. The cumulative property is filled during the dfs
+   exploration of the graph: if not explored the number is 0.
+
+   Each vertex has a property (vertex_name_t) that holds a const pointer to the
+   SimTrack that originated that vertex and the cumulative number of simHits of
+   all its outgoing edges. The cumulative property is filled during the dfs
+   exploration of the graph: if not explored the number is 0.
+
+   Stable particles are recovered/added in a second iterations and are linked
+   to ghost vertices with an offset starting from the highest generated vertex.
+
+   Multiple decays of a single particle that retains its original trackId are
+   merged into one unique vertex (the first encountered) in order to avoid
+   multiple counting of its associated simHits (if any).
+
+*/
+struct EdgeProperty {
```

Comment:
Why are some structs defined outside of the `CaloTruthAccumulator` class and others defined within it?

Reply:
I think that the graph structs are more general and could stay outside. In the end, everything is contained in the cc file that is, itself, a plugin, so there should be no pollution outside of this code.